### PR TITLE
refactor!: Build DD4hep geometry from hierarchy, use VariantParameters

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR changes the DD4hep integration model. Instead of the DD4hep factory code having to depend on ACTS or the ActsDD4hep glue lib, they now use a mechanism we merged into DD4hep (as of `1.21`): [`MapStringVariantStruct`](https://dd4hep.web.cern.ch/dd4hep/reference/structdd4hep_1_1rec_1_1MapStringVariantStruct.html), which is wrapped into a detector element *extension* as `VariantParameters`.

This extension is independent of ACTS and encodes key value pairs, with the keys being strings and the values being any of string, int, float or bool. ACTS will use information attached to detector elements via this mechanism to identify layers, surface axes and material properties (layer and boundary).

This PR also removes the need to active subdetectors to be marked up using a specific extension: using the DD4hep `DetType` enum, ACTS will look for detector elements flagged `TRACKER` and `BARREL`, `ENDCAP` or `BEAMPIPE` and handle them accordingly. Passive structures that cannot be flagged with the above use a *variant* parameter `passive_layer` to tell ACTS to translate them.

## Possible parameters ACTS will interpret

ACTS geometry translation uses parameters attached to DD4hep detector elements via the `VariantParameters` extension. You can set this extension from your factory code, or use DD4hep's plugin mechanism to set parameters on detector elements via name. Here's an excerpt from the Open Data Detector on how to use this to mark a passive structure as a *passive layer*:

```xml
<lccdd>
  <detectors>
    <detector id="ODD_Solenoid_ID" name="Solenoid" type="ODDCylinder" beampipe="false" vis="Aluminum">
      <type_flags type="DetType_TRACKER" />
      <boundary_material surface="inner" binning="binPhi,binZ" bins0="mat_sol_bPhi" bins1="mat_sol_bZ"/>
      <tubs name="Solenoid" rmin="sol_rmin" rmax="sol_rmax" dz="sol_hlength" material="Aluminum">
        <layer_material surface="representing" binning="binPhi,binZ" bins0="mat_sol_bPhi" bins1="mat_sol_bZ"/>
      </tubs>
    </detector>
  </detectors>

  <plugins>
    <plugin name="DD4hep_ParametersPlugin">
      <argument value="Solenoid"/>
      <argument value="passive_layer: bool = true"/>
    </plugin>	
  </plugins>
</lccdd>
```

* Layer

  * `envelope_{r,z}_{min,max}`: explicit envelope for a layer
  * `layer_material`: mark a layer as passive (use this if you want to add a passive layer that is not a beampipe)
  * Surface binning:

    * `surface_binning`: set to true to indicate that explicit surface binning is set.
    * `surface_binning_n_{phi,r}`: surface binning for a layer

  * Layer material:

    * `layer_material`: set to true to indicate that the layer has layer material configuration

      * `layer_material_representing`: set to true to indicate representing layer material
      * `layer_material_inner`: set to true to indicate inner layer material
      * `layer_material_outer`: set to true to indicate outer layer material
      * `layer_material_{representing,inner,outer}_bin{X,Y,Z,R,Phi,RPhi,H,Eta,Mag}` to give the number of bins in a direction

* Sensor

  * `axis_definitions`: local axis definitions for a sensor. Default: `XYZ`. See {class}`Acts::DD4hepDetectorElement` for details

* Volume / subdetector

  * `boundary_material`: set to true to indicate boundary material is set
  * `boundary_material_{negative,positive,inner,outer}` to indicate which boundary material surfaces should be set
  * `boundary_material_{negative,positive,inner,outer}_bin{X,Y,Z,R,Phi,RPhi,H,Eta,Mag}` to give the number of bins in a direction

### CODE REVIEW

**Comment by paulgessinger** [comment]
Ok, so the python level tests and ROOT hashes seem to be happy, as is the physmon job.

I think this means that the geometry is identical with this PR to main at least as far as the test jobs can tell.

The remaining failing jobs are compilation failures because the DD4hep version in them doesn't have the required changes.

**Comment by paulgessinger** [comment]
DD4hep tag [`v1.21`](https://github.com/AIDASoft/DD4hep/releases/tag/v01-21) exists now, I'll update this PR with new images.

**Comment by paulgessinger** [comment]
I think this is good to go. I've updated the documentation to discuss the new integration model. Let me know if that is not clear enough or needs more detail.

Steps to land this change:
1. I've created a pre-release tag on ODD for the WIP version of the changes needed on that end: [`v3.0.0-rc.1`](https://gitlab.cern.ch/acts/OpenDataDetector/-/tags/v3.0.0-rc.1) so we retain the ODD commit this branch currently points at. 
2. I will now create a pre-release tag on the current HEAD of my branch `v20.0.0-rc.1`. I will use that to update the corresponding [ODD MR](https://gitlab.cern.ch/acts/OpenDataDetector/-/merge_requests/31) with an ACTS version that can run the ODD CI (and won't be garbage collected).
3. Once the [ODD MR](https://gitlab.cern.ch/acts/OpenDataDetector/-/merge_requests/31) is merged, I'll bump this PR to the updated `v3.0.0` ODD tag, and it's good to go.

Any thoughts? (@asalzburger, @andiwand)

**Review by andiwand — APPROVED** [review]
7k lines gone - that's what I call a proper refactor :smile: 

sorry I browsed through the code but could not come up with any meaningful comments

**Comment by paulgessinger** [comment]
Ha this is mostly because of some files I deleted. I moved this to #1337. Once that's merged this PR will shrink a bit.

File: Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp
Comment (asalzburger) [inline Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp:12]: Does this need an `#include <exception>`

File: Plugins/DD4hep/src/ConvertDD4hepDetector.cpp
Comment (asalzburger) [inline Plugins/DD4hep/src/ConvertDD4hepDetector.cpp:211]: We should probably specify at some point that this is a cylindrical detector converter with barrel/EC structure.

File: Plugins/DD4hep/src/ConvertDD4hepDetector.cpp
Comment (asalzburger) [inline Plugins/DD4hep/src/ConvertDD4hepDetector.cpp:619]: Not sure why we hadn't used our own naming conventions here.

File: docs/examples/howto/howto.rst
Comment (asalzburger) [inline docs/examples/howto/howto.rst:11]: This is probably not fully related that the renaming comes here, but that's ok

**Review by asalzburger — APPROVED** [review]

File: Core/src/Geometry/CylinderVolumeBuilder.cpp
Comment (paulgessinger) [reply Core/src/Geometry/CylinderVolumeBuilder.cpp:169]: Yeah. This is lumped in with this PR. Can remove it if you prefer.

File: Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp
Comment (paulgessinger) [reply Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp:12]: No this actually needs to go away again. Good catch.

File: Plugins/DD4hep/src/ConvertDD4hepDetector.cpp
Comment (paulgessinger) [reply Plugins/DD4hep/src/ConvertDD4hepDetector.cpp:211]: Absolutely. I think the documentation is pretty clear about that already though.

File: docs/examples/howto/howto.rst
Comment (paulgessinger) [reply docs/examples/howto/howto.rst:11]: I can remove this.

File: Core/src/Geometry/CylinderVolumeBuilder.cpp
Comment (paulgessinger) [reply Core/src/Geometry/CylinderVolumeBuilder.cpp:169]: This currently runs this always. I can put this behind a log check, but it's the geometry construction so I don't think it matters.

File: Plugins/DD4hep/src/ConvertDD4hepDetector.cpp
Comment (paulgessinger) [reply Plugins/DD4hep/src/ConvertDD4hepDetector.cpp:619]: 🤷

File: Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp
Comment (paulgessinger) [reply Plugins/DD4hep/include/Acts/Plugins/DD4hep/DD4hepConversionHelpers.hpp]: Should be gone.

**Comment by paulgessinger** [comment]
Ok, in the standalone ODD CI I get histogram differences. Marking this WIP again until I've investigated.

**Comment by paulgessinger** [comment]
Differences are understood to have been related to #1298.

**Comment by paulgessinger** [comment]
https://gitlab.cern.ch/acts/OpenDataDetector/-/merge_requests/31 should be merged first.

**Comment by asalzburger** [comment]
Paul, there are conflicts here, can you update - otherwise you addressed my comments.

**Comment by paulgessinger** [comment]
@asalzburger this now points at ODD `v3.0.0` and is ready to go. Can you approve?

**Comment by asalzburger** [comment]
Updated branch, and will approve when done.

**Pull request merged**
